clang/lib/CodeGen/CGVTables.cpp

```diff
@@ -1185,6 +1190,21 @@ bool CodeGenVTables::isVTableExternal(const CXXRecordDecl *RD) {
       TSK == TSK_ExplicitInstantiationDefinition)
     return false;
 
+  // Itanium C++ ABI [5.2.3]:
+  // Virtual tables for dynamic classes are emitted as follows:
+  //
+  // - If the class is templated, the tables are emitted in every object that
+  // references any of them.
+  // - Otherwise, if the class is attached to a module, the tables are uniquely
+  // emitted in the object for the module unit in which it is defined.
+  // - Otherwise, if the class has a key function (see below), the tables are
+  // emitted in the object for the translation unit containing the definition of
+  // the key function. This is unique if the key function is not inline.
+  // - Otherwise, the tables are emitted in every object that references any of
+  // them.
+  if (RD->isInNamedModule())
+    return RD->shouldEmitInExternalSource();
```

Comment:
Shouldn't need to test "isInNamedModule" here, though? Would it be adequate to do:
```
if (RD->shouldEmitInExternalSource())
  return true;
```

Ah, I geuss you want to return false ASAP too - I guess looking at some of the implementation details of `shouldEmitInExternalSource` and using those APIs directly here - Always -> true, Never -> false, I think?

Like `adjustGVALinkageForExternalDefinitionKind` already does? Or even reusing some of that functionality somehow.

Reply:
I tried. But it can't work the case we compile the AST directly to the object file. In this case, `hasExternalDefinitions` returns EK_ReplyHazy, but we need to return false directly here. I think it looks good now. If we want to extend this to header modules with object files, it will be easy.

Reply:
That seems unfortunate - having these things work through different codepaths/states when compiling directly rather than via a pcm. It'd be good to make those paths converge (making `hasExternalDefinitions` provide the same answers/do the right thing when compiling directly without the intermediate pcm).

Reply:
I think we would have to maintain both the one phase compilation and two phase compilation path. Maybe we can make it better by refining the API.